src/main/java/nl/tudelft/lifetiles/graph/controller/GraphController.java

```diff
@@ -213,45 +253,88 @@ private void repaint() {
 
-            root = new Group();
-
             wrapper.hvalueProperty().addListener(
                     (observable, oldValue, newValue) -> {
-                        repaintPosition(root, newValue.doubleValue());
+                        repaintPosition(newValue.doubleValue());
                     });
 
-            Rectangle clip = new Rectangle(getMaxUnifiedEnd(graph)
-                    * VertexView.HORIZONTALSCALE, 0);
-            root.getChildren().add(clip);
-
-            repaintPosition(root, wrapper.hvalueProperty().doubleValue());
+            repaintPosition(wrapper.hvalueProperty().doubleValue());
         }
     }
 
     /**
      * Repaints the view indicated by the bucket in the given position.
      *
-     * @param root
-     *            Root group used to store the visualized graph in.
      * @param position
      *            Position in the scrollPane.
      */
-    private void repaintPosition(final Group root, final double position) {
-        int nextPosition = getBucketPosition(position);
-        if (currentPosition != nextPosition || forceRepaintPosition) {
+    private void repaintPosition(final double position) {
+        double graphWidth = getMaxUnifiedEnd(graph)
+                * VertexView.HORIZONTALSCALE * scale;
+
+        double screenWidth = wrapper.getViewportBounds().getWidth();
+        double diffWidth = graphWidth - screenWidth;
+        double scaledScreenWidth = screenWidth * 2 * scale;
+
+        double newPosition = (position * (screenWidth / 2 + diffWidth))
+                / graphWidth;
+
+        double start = ((newPosition * graphWidth - scaledScreenWidth) / scale)
+                / VertexView.HORIZONTALSCALE;
+
+        double end = ((newPosition * graphWidth + scaledScreenWidth) / scale)
+                / VertexView.HORIZONTALSCALE;
+
+        int startBucket = getStartBucketPosition(start);
+        int endBucket = getEndBucketPosition(end) + 1;
+
+        if (currEndPosition != endBucket && currStartPosition != startBucket
+                || repaintNow) {
             if (graphNode != null) {
                 graphNode.getChildren().clear();
             }
             graphNode = new Group();
-            graphNode.getChildren().add(drawGraph(nextPosition));
-            root.getChildren().add(graphNode);
-            wrapper.setContent(root);
-            currentPosition = nextPosition;
-            forceRepaintPosition = false;
+
+            graphNode.getChildren().add(drawGraph(startBucket, endBucket));
+
+            Rectangle clip = new Rectangle(graphWidth, 0);
+
+            Group newRoot = new Group();
+            newRoot.getChildren().add(clip);
+            newRoot.getChildren().add(graphNode);
+
+            wrapper.setContent(newRoot);
+
+            currEndPosition = endBucket;
+            currStartPosition = startBucket;
+
+            repaintNow = false;
         }
     }
 
     /**
+     * Zoom in on the current graph.
+     */
+    private void zoomIn() {
```

Comment:
duplicate code; `zoomIn()` and `zoomOut()` should use some private `zoom()` method with a zoomfactor as argument.